src/simularium/VisDataCache.ts

```diff
@@ -0,0 +1,247 @@
+        cacheEnabled?: boolean;
+    }): void {
+        const { maxSize, cacheEnabled } = options;
+        if (cacheEnabled !== undefined) {
+            this.cacheEnabled = cacheEnabled;
+        }
+        if (maxSize !== undefined) {
+            this.maxSize = maxSize;
+            this.cacheSizeLimited = maxSize > 0;
+        }
+    }
+
+    private frameAccessError(msg?: string): CachedFrame {
+        throw new FrontEndError(
+            `Error accessing frame: ${msg}`,
+            ErrorLevel.WARNING
+        );
+    }
+
+    public hasFrames(): boolean {
+        return this.numFrames > 0 && this.head !== null && this.tail !== null;
+    }
+
+    /**
+     * Walks the cache looking for node that satisfies condition
+     * returns the node if found, otherwise returns null,
+     * starts at head if firstNode is not provided.
+     */
+    private walkLinkedList(
+        condition: (data: LinkedListNode) => boolean,
+        firstNode?: LinkedListNode
+    ): LinkedListNode | null {
+        let currentNode = firstNode || this.head;
+        while (currentNode) {
+            if (condition(currentNode)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    public containsTime(time: number): boolean {
+        if (time < this.getFirstFrameTime() || time > this.getLastFrameTime()) {
+            return false;
+        }
+        return !!this.walkLinkedList((node) => node.data.time === time);
+    }
+
+    public containsFrameAtFrameNumber(frameNumber: number): boolean {
+        if (
+            frameNumber < this.getFirstFrameNumber() ||
+            frameNumber > this.getLastFrameNumber()
+        ) {
+            return false;
+        }
+        return !!this.walkLinkedList(
+            (node) => node.data.frameNumber === frameNumber
+        );
+    }
+
+    public getFirstFrame(): CachedFrame {
+        if (!this.head) {
+            return this.frameAccessError("No data in cache.");
+        }
+        return this.head.data;
+    }
+
+    public getFirstFrameNumber(): number {
+        return this.head?.data.frameNumber || -1;
+    }
+
+    public getFirstFrameTime(): number {
+        return this.head?.data.time || -1;
+    }
+
+    public getLastFrame(): CachedFrame {
+        if (!this.tail) {
+            return this.frameAccessError(" No data in cache.");
+        }
+        return this.tail.data;
+    }
+
+    public getLastFrameNumber(): number {
+        return this.tail?.data.frameNumber || -1;
+    }
+
+    public getLastFrameTime(): number {
+        return this.tail?.data.time || -1;
+    }
+
+    private getFrameAtTimeOrFrameNumber(
+        condition: "time" | "frameNumber",
+        value: number
+    ): CachedFrame {
+        if (!this.head) {
+            return this.frameAccessError("No data in cache.");
+        }
+        const frame = this.walkLinkedList(
+            (node) => node.data[condition] === value
```

Comment:
See questions above...

Reply:
Ok I am making these less tightly bound to property names, `getFrameAtTime` and `getFrameAtFrameNumber` will provide conditional checks instead of strings, and `getFrameAtTimeOrFrameNumber` becomes `getFrameAtCondition`, it does error handling mostly. 

For the time check I am reinstating `compareTimes` but with a zero time step, I still think this a) matches what was in the code before b) is not functionally different than a straight equality.

Not giving up on this question though I've reached out to @ascibisz and adding this to lest of next steps, if we need a robust time check here I will follow up to ensure that `compareTimes` gets the proper time step, considering that out of scope for now.